governance/default_test.go

```diff
@@ -97,6 +97,12 @@ var tstData = []voteValue{
 	{k: "governance.unitprice", v: float64(-10), e: false},
 	{k: "governance.unitprice", v: "25000000000", e: false},
 	{k: "governance.unitprice", v: true, e: false},
+	{k: "governance.deriveshaimpl", v: float64(0.0), e: true},
+	{k: "governance.deriveshaimpl", v: float64(0.1), e: false},
+	{k: "governance.deriveshaimpl", v: uint64(2), e: true},
+	{k: "governance.deriveshaimpl", v: float64(-1), e: false},
+	{k: "governance.deriveshaimpl", v: "2", e: false},
+	{k: "governance.deriveshaimpl", v: true, e: false},
```

Comment:
What happen if we vote 3 which doesn't exist for the deriveSha?

Reply:
In that case, vote passes and make it to the governance database.
However, below if statement makes it to use DefaultDeriveShaImpl = 0 = Original.

```go
	if _, ok := impls[implType]; ok {
		return implType
	} else {
blukat29 marked this conversation as resolved.
Show resolved
		logger.Error("Unrecognized deriveShaImpl, falling back to default impl", "impl", implType)
		return int(params.DefaultDeriveShaImpl)
	}
```

Reply:
@blukat29 IMO, it might be better to block for invalid values for unexpected situation.

Reply:
Currently, there are multiple inflows of DeriveShaImpl
- genesis.json
- ChainConfig DB
- `governance.vote()` API argument
- `block.voteData`
- `block.governanceData`
- `GovParam.sol:getParams()`

Unless all inflows are properly guarded, the fallback logic must be there to prevent deadlock situation.
e.g. an ancient genesis block contained impl=3.

Thanks for the suggestion though.
I am now designing how to centralize all parameter-specific logic into one location. Hopefully in next version, I can provide more visibility to how parameters are processed throughout the Klaytn code.